Refactor recording of CLI subcommand invocations

happz opened the pull request:
To store invocations of various subcommands and their effects on classes
like `Test` or `Cleanup`, `CLIInvocation`, new dataclass, has been
added. PR attempts to sort out the processing of these invocations, document
places that record them, prevent accidental overwriting of invocations,
and support multiple records.

On top of this, experimental support for multiple command line phases
is added, with `--insert` and `--update` options for tweaking phases
read from fmf data.

happz commented:
@janhavlin So, this is the first attempt: I realized I'm still very puzzled about many aspects, so I focused on giving things better names I could grasp and follow.

* there's a class called `CLIInvocation`, for each mention of a "class" on the command line - `run`, `plan`, `test`, `report`, ... - there would be one `CLIInvocation` instance. It bundles together a Click context & options given on the command line. Options can be derived from the context, but some plugins do nasty things with options, so let's store them aside as well.
* then I tried to separate classes derived from `Common` into two groups: classes than can be mentioned on the command line just once, and classes that can - and are expected to be in the future - mentioned multiple times. The first group - `Run`, `Plan`, maybe also `Clean`, there's just a single `run` in `tmt run ...`, and just a single `plan` in `tmt run ... plan --name foo` and so on; but we plan to support multiple invocations of steps, like `report -h html ... report -h display`, or multiple invocations of `login` pseudo-step.

`Common` class now has a singular `cli_invocation` attribute, there's just a single chance to set it, and it's forbidden to change it. The other group behaves in the similar way, but instead of raising an exception, a warning is printed out when a step is mentioned multiple times on the command line.

From this, I hope I could follow the flow of "CLI invocations" of various components, add checks and debug messages to better split between options passed to a `report` step and `report` plugins in command lines like `tmt run -a report -v report -h html -vvv report -h display -vv` - this should help us then convert *some* of these sequences of options into plugin instances. PR is probably broken, and unit tests failed (although `tmt -c trigger=commit -vv run -a report -h display report -h html plan -n '^/plans/features/core$' t -n 'unit'` finished successfully), but feel free to ask questions and raise comments.

cc @psss

happz commented:
For experiments, the following helped me debug the basic issues:

```shell
$ tmt -c trigger=commit -vv \
      run -a \
      report --insert -h display --order 70 \
      report --insert -h html \
      plan -n '^/plans/features/core$' \
      test -n '/tests/core/ls'
```

janhavlin commented:
```
$ tmt -c trigger=commit -vv \
      run -a \
      report --insert -h display --order 70 \
      report --insert -h html \
      plan -n '^/plans/features/core$' \
      test -n '/tests/core/ls'
/var/tmp/tmt/run-048
Found 1 plan.

/plans/features/core
summary: Verify core functionality

plan failed

The exception was caused by 1 earlier exceptions

Cause number 1:

    '<' not supported between instances of 'NoneType' and 'int'
```
Looks unexpected. Any idea? Haven't debugged it yet.

happz commented:
> ```
> $ tmt -c trigger=commit -vv \
>       run -a \
>       report --insert -h display --order 70 \
>       report --insert -h html \
>       plan -n '^/plans/features/core$' \
>       test -n '/tests/core/ls'
> /var/tmp/tmt/run-048
> Found 1 plan.
> 
> /plans/features/core
> summary: Verify core functionality
> 
> plan failed
> 
> The exception was caused by 1 earlier exceptions
> 
> Cause number 1:
> 
>     '<' not supported between instances of 'NoneType' and 'int'
> ```
> 
> 
>     
>   
> 
> Looks unexpected. Any idea? Haven't debugged it yet.

Probably comparison of `order` fields, but `order` should have a default value, therefone no ` one` is allowed to sneak in. If you add `—order` to the second phase, would the issue go away? Check where the option is defined, there should be some related changes in this PR.

happz commented:
> > ```
> > $ tmt -c trigger=commit -vv \
> >       run -a \
> >       report --insert -h display --order 70 \
> >       report --insert -h html \
> >       plan -n '^/plans/features/core$' \
> >       test -n '/tests/core/ls'
> > /var/tmp/tmt/run-048
> > Found 1 plan.
> > 
> > /plans/features/core
> > summary: Verify core functionality
> > 
> > plan failed
> > 
> > The exception was caused by 1 earlier exceptions
> > 
> > Cause number 1:
> > 
> >     '<' not supported between instances of 'NoneType' and 'int'
> > ```
> >   
> > Looks unexpected. Any idea? Haven't debugged it yet.
> 
> Probably comparison of `order` fields, but `order` should have a default value, therefore no `None` is allowed to sneak in. If you add `—order` to the second phase, would the issue go away? Check where the option is defined, there should be some related changes in this PR.

Ok, this was the problem:

> but `order` should have a default value

There was no default, that was the issue. Fixed.

janhavlin commented:
Yeah, it passes now for me.

happz commented:
I think I can explain both, to some extent.

> I guess it has to do something with defaults:

"default phases" is probably more fitting - when there are no fmf data for a step, empty one with the default `how` is constructed in `Step.__init__()`. Hence why there's *one* of `display` phases. Now...

> 
> ```yaml
> /plan:
>     execute:
>         how: tmt
>     discover:
>         how: shell
>         tests:
>         - name: foo
>           test: echo 
>     provision:
>         how: local
> ```
> 
> See that 'display' is twice, verbosity is not respected and report -h html is missing as well:
> 
> ```
> $ tmt run -a report  -h display -vvv report --insert -h html
>     report
>         how: display
>         how: display
>         summary: 1 test passed
> ```
> 
> As if the 'report --insert' got added to the plan first and just after that the 'replace' part of CLI got evaluated.

Correct observation, that's exactly how it's implemented in this PR. I'm not saying it's the best way, it's just a prototype, I'm open to suggestions.

You already know this, but bear with me, please, it's all tangled up, and verbose explanations help me follow the trail: the current tmt behavior is, when a step is mentioned on the command line, it "converts" all phases from fmf to itself. If the phase had the same `how` as the command line input, good, it's updated by the command line input, if `how` was different, it's replaced with the command line input:

```python
                if raw_datum['how'] == how:
                    self.debug(f'  compatible step data:   {raw_datum}', level=4)
                    _raw_data.append(raw_datum)

                # Mismatch, throwing away, replacing with new `how` - but we can keep the name.
                else:
                    self.debug(f'  incompatible step data: {raw_datum}', level=4)
                    _raw_data.append({
                        'name': raw_datum['name'],
                        'how': how
                        })
```

So, there's this default `display` from "no `report` in fmf tree" beginning, then CLI applies:

* `report -h display -vvv` "overwrites" this default phase, let's ignore the verbosity for now, but nothing changes that much. But: since the current behavior has no options like `--insert`, and phase from CLI overwrites *every single existing phase*, and the last CLI phase wins, CLI input without `--insert` or `--update` are indeed put aside to be evaluated later.
* `report --insert -h html` - yep, inserts `html` report phase.
* then the previous `display` is re-evaluated and overwrites your precious `html` phase, and you get two `display` phases and the behavior you observed.

Not perfect, but somehow compatible with the current behavior. I guess the question is, do we care about the current behavior, how should these "option-less" phases interact with those with `--insert` or `--update`? What is the expected outcome of `tmt run report -h html report -h reportportal report --insert -h html`, for example? Would there be an implicit `--insert`, `--update`, `--foo` action?


> When I do both 'report' as --insert it executes html but display is twice again (and verbosity not respected neither):
> 
> ```
> $ tmt run -a report --insert -h display -vvv report --insert -h html
>     report
>         how: display
>         how: display
>         how: html
>         output: /var/tmp/tmt/run-043/demo/plan/report/default-2/index.html
>         summary: 1 test passed
> ```

Putting the verbosity aside for now, the default phase from "no `report` in fmf, let's spawn a default phase then" step is the one extra, both CLI phases had `--insert`, no overwrites, no changes, just additions.

And the verbosity... Yeah, that should be honored, no idea what's happening there, I'll debug it.

happz commented:
Right, the verbosity issue is caused by the fight between plugin inputs: there are "step data", basically a bundle of plugin-specific keys, filled by fmf or CLI, but there are also inputs that are *not* saved in this structure, like `--verbose` or `--debug`. And when a plugin instance is instantiated, only the step data are given to the plugin class, while CLI invocation with `--verbose` is taken from a class variable. Hence different phases of the same plugin can't have different verbosity. Not good, but at least it's a problem I now understand.

I suppose we need to tie together step data bundles with invocations that affected them. Either by collecting `CLIInvocation` objects and attaching them to step data bundles or by moving keys like `verbose` or `debug` to step data - and I would rather not do that because that would make them appear like something one can specify in fmf.

thrix commented on tests/__init__.py:
```diff
@@ -0,0 +1,37 @@
+import click.testing
+
+
+def reset_common() -> None:
```
could we add info into this file, or something, so it is cleared why this was added please?

psss replied:
Covered.

thrix reviewed (changes requested):
not finished with review yet :( just firs small ask

sopos commented:
well, it work only if I provide `--how` on the cmd line.
```yaml
provision:
  - how: minute
    image: fedora
    role: server
    name: server
  - how: minute
    image: fedora
    role: client
    name: client
  - how: minute
    image: fedora
    role: controller
    name: controller

# tmt run -vvvddd provision --update --name server --image fedora-37
Usage: tmt run provision [OPTIONS]
Try 'tmt run provision --help' for help.

Error: No such option: --image Did you mean --name?


# tmt run -vvvddd provision --update --how minute --name server --image fedora-37
Using tree '/home/dapospis/test/git/beakerlib/ControlFlow'. 
/var/tmp/tmt/run-033
tmt version: 1.27.dev (da985a2d)
tmt command line: /usr/bin/tmt run -vvvddd provision --update --how minute --name server --image fedora-37
Read file '/var/tmp/tmt/run-033/run.yaml'.
Run data not found.
Enabled steps: provision

```

Two things:
1. I would expect to update all phases if the `--name` was not provided
2. omitting `--how` should be fine as I defined it in the plan

psss commented on tmt/utils.py:
```diff
@@ -5047,17 +5140,24 @@ def _load_keys(
                 # try+except seems to work better than get(), especially when
                 # semantic of fmf.Tree.get() is slightly different than that
                 # of dict().get().
```
The `try-except` clause has been removed. Please, update the comment as well. Would be also nice to shortly explain the difference between `source_keyname` and `source_keyname_cli`.

happz replied:
Fixed in 93b996cb

psss reviewed (changes requested):
Thanks a lot for implementing this! Looks very good and works nicely. Added a couple of comments.

psss commented:
> Perfect is the enemy of good, progress, not perfection, blah blah blah, it's simply so complicated I wouldn't touch it with a ten-feet pole **now**. I like starting small, shaping the field for future extensions, cleaning, documenting...

Agreed, let's get started with something working, polishing will come later. This code area is really complicated.

happz commented on tests/__init__.py:
```diff
@@ -0,0 +1,50 @@
+        klass.cli_invocation = None
+
+
+class CLIRunner(click.testing.CliRunner):
```
Fixed in 93b996cb

happz commented on tests/core/phases/main.fmf:
```diff
@@ -0,0 +1,3 @@
+summary: Verify fmf and CLI phases play nicely together
+test: ./test.sh
+framework: beakerlib
```
Fixed in 93b996cb

happz commented on tests/core/phases/test.sh:
```diff
@@ -0,0 +1,155 @@
+            report --update --name default-0 -h html"
+
+        check "with-order" "A single 'html' phase with preserved order shall exist" "default-0:html:60"
+    rlPhaseEnd
```
Added a check in 93b996cb

happz commented on tmt/steps/__init__.py:
```diff
@@ -62,6 +74,64 @@
+    option(
+        '--name',
+        type=str,
+        help='Which existing phase should be updated when --update is used.'),
```
Fixed in 93b996cb

happz commented on tmt/steps/__init__.py:
```diff
@@ -62,6 +74,64 @@
+    ])
+
+
+class DefaultNameGenerator:
```
Added in 93b996cb

happz commented on tmt/steps/__init__.py:
```diff
@@ -62,6 +74,64 @@
+
+        return DefaultNameGenerator(actual_name_keys)
+
+    def restart(self) -> None:
```
`reset` was necessary, but it does not seem to be needed or used anymore. I removed it in 93b996cb. And added plenty of docstrings everywhere around.

happz commented on tmt/steps/__init__.py:
```diff
@@ -102,6 +172,13 @@ def is_in_standalone_mode(self) -> bool:
+    'how': str,
+    'name': str,
+    'insert': str,
+    'update': str
```
The whole type is gone in 93b996cb

happz commented on tmt/steps/__init__.py:
```diff
@@ -456,52 +539,136 @@ def wake(self) -> None:
+        # materialize when it's going to be thrown away anyway.
+        debug = functools.partial(self.debug, level=4, topic=tmt.log.Topic.CLI_INVOCATIONS)
+
+        debug('updating phases by CLI invocations')
```
Fixed in 93b996cb

happz commented on tmt/steps/__init__.py:
```diff
@@ -456,52 +539,136 @@ def wake(self) -> None:
+
+            return raw_datum
+
+        def _patch_raw_datum(raw_datum: _RawStepData, incoming_raw_datum: _RawStepData) -> None:
```
Fixed in 93b996cb

happz commented on tmt/steps/__init__.py:
```diff
@@ -456,52 +539,136 @@ def wake(self) -> None:
+
+                postponed_invocations.append(invocation)
+
+        for i, invocation in enumerate(postponed_invocations):
```
Added in 93b996cb, comments above and below this line.

happz commented on tmt/steps/__init__.py:
```diff
@@ -1092,7 +1259,7 @@ def wake(self) -> None:
         #         f'current data is {self.data}'
 
-        self._update_data_from_options()
+        # self._update_data_from_options()
```
Yes, the whole method is gone in 93b996cb

happz commented on tmt/utils.py:
```diff
@@ -894,65 +879,115 @@ def __str__(self) -> str:
-           The given context will overwrite any previously saved context.
+        if cls.cli_invocation is not None:
+            raise GeneralError(
+                f"{cls.__name__} attempted to save second CLI context: {cls.cli_invocation}")
```
Fixed in 93b996cb

happz commented on tmt/utils.py:
```diff
@@ -894,65 +879,115 @@ def __str__(self) -> str:
+        CLI invocation attached to this instance or its parents.
 
-    def _save_cli_context_to_instance(self, context: 'tmt.cli.Context') -> None:
+        :returns: instance-level CLI invocation, or, it there is none,
```
Fixed in 93b996cb

psss reviewed (commented):
Thanks for addressing all comments. Now looks good, however tests seem to be red.

happz commented:
> Thanks for addressing all comments. Now looks good, however tests seem to be red.

@psss tests should be fixed with https://github.com/teemtee/tests/pull/4/, which is failing right now, too, the chicken and egg problem. It's the `CliRunner` situation.

Although I just noticed the internal branch of tests is failing as well, I have to fix that one, no PR there yet.

psss commented:
Thanks, both merged, let's see.

psss commented:
/packit test

psss commented on tmt/cli.py:
```diff
@@ -86,6 +86,34 @@ class Context(click.Context):
 
 
+@dataclasses.dataclass
+class CLIInvocation:
```
```suggestion
class CliInvocation:
```

The last remaining naming inconsistency.

happz replied:
Renamed in 20d22859.

FTR, camel case sucks :)

psss reviewed (approved):
Tests are red but it seems that's an unrelated documentation issue. So, I guess, we're ready to merge? Or, @thrix, do you still want to finish your review?

martinhoyer commented:
I assume you're all aware, but tmt's plans/remote/polarion is using the teemtee/tests repo, so merging https://github.com/teemtee/tests/commit/f7cac2f1f1dbb6e56cb7406f6abb74cdefc3278c there without merging this PR broke tests here:

https://artifacts.dev.testing-farm.io/d99d6565-acf1-42ba-b63f-e1f1f912690c/

thrix reviewed (changes requested):
Some small comments from me

happz commented on tests/core/phases/test.sh:
```diff
@@ -0,0 +1,162 @@
+
+    rlPhaseStartTest "Test /with-multiple-reports with --update"
+        rlRun -s "$run plan -n /with-multiple-reports \
+            report --update --name default-0 -h html --display-guest never"
```
Probably not in this test, but in general, yes, I can't spot a test that would explicitly cover `--update` with `--how` overriding the fmf input. I'll add one.

thrix replied:
thx

happz replied:
Added in bbd6321b

happz commented on tmt/steps/__init__.py:
```diff
@@ -62,6 +74,85 @@
+    option(
+        '--name',
+        type=str,
+        help="Name of the existing phase which should be updated when '--update' is used."),
```
I'm afraid I don't follow, can you provide an example of the situation and what you would like to see happen?

thrix replied:
I have a plan

```
❯ cat plans/example.fmf
summary: Basic smoke test
discover:
    how: fmf
execute:
    how: tmt
report:
  - name: html
    how: html
  - how: display
```

I want to update the second report phase with `how: display`.
It is hard to find out the `name` attribute of it.
It is not displayed anywhere, the only place I was able to spot it was when running with `-ddd`:

```
❯ tmt plan show -ddd |& grep default-0
    DiscoverPlugin.delegate(step=discover, data=None, raw_data={'how': 'fmf', 'name': 'default-0'})
    DiscoverPlugin.delegate(step=discover, data=DiscoverFmfStepData(name='default-0', how='fmf', order=50, summary=None, where=[], dist_git_source=False, dist_git_type=None, url=None, ref=None, path=None, test=[], link=[], filter=[], exclude=[], modified_only=False, modified_url=None, modified_ref=None, dist_git_init=False, dist_git_remove_fmf_root=False, dist_git_merge=False, dist_git_extract=None, sync_repo=False, fmf_id=False, repository=None, revision=None), raw_data=None)
    ProvisionPlugin.delegate(step=provision, data=None, raw_data={'name': 'default-0', 'how': 'virtual'})
    ProvisionPlugin.delegate(step=provision, data=ProvisionTestcloudData(name='default-0', how='virtual', order=50, summary=None, hardware=None, role=None, guest=None, facts=GuestFacts(in_sync=False, arch=None, distro=None, kernel_release=None, package_manager=None, has_selinux=None, is_superuser=None, os_release_content={}, lsb_release_content={}), port=None, user='root', key=[], password=None, ssh_option=[], image='fedora', memory=2048, disk=40, connection='session', arch='x86_64', list_local_images=False, image_url=None, instance_name=None), raw_data=None)
    PreparePlugin.delegate(step=prepare, data=None, raw_data={'name': 'default-0', 'how': 'shell'})
    PreparePlugin.delegate(step=prepare, data=PrepareShellData(name='default-0', how='shell', order=50, summary=None, where=[], script=[]), raw_data=None)
    ExecutePlugin.delegate(step=execute, data=None, raw_data={'how': 'tmt', 'name': 'default-0'})
    ExecutePlugin.delegate(step=execute, data=ExecuteInternalData(name='default-0', how='tmt', order=50, summary=None, where=[], duration='1h', exit_first=False, script=[], interactive=False, no_progress_bar=False), raw_data=None)
    ReportPlugin.delegate(step=report, data=None, raw_data={'how': 'display', 'name': 'default-0'})
    ReportPlugin.delegate(step=report, data=ReportDisplayData(name='default-0', how='display', order=50, summary=None, display_guest='auto'), raw_data=None)
    FinishPlugin.delegate(step=finish, data=None, raw_data={'name': 'default-0', 'how': 'shell'})
    FinishPlugin.delegate(step=finish, data=FinishShellData(name='default-0', how='shell', order=50, summary=None, where=[], script=[]), raw_data=None)
```

While I guess this problem was not introduced in this MR, it is making it more visible, as I want to use the `--name` attribute to update a specific phase.

I would like to propose, in case of multiple phases for the step, display the name always, including the default one.
That would make this easy.

happz replied:
Ah, now I understand :) Well, that's on purpose, actually, tmt will actively *hide* names looking like a default name. That's why `tmt plan show` does not print it out. It was discussed on IRC - I should be able to find the discussion if needed -, and it was a feature @psss wanted to keep around: https://github.com/teemtee/tmt/pull/1384#discussion_r945656170.

Showing the `name` key always, i.e. not "hiding" info from the user as I described it a year ago, would fit my nature. But, since @psss felts strongly about this, I'd not block this PR on changing this particular behavior, I'd rather do so in a separate patch to give @psss a chance to voice his concerns, although I believe the new `--update` requiring the knowledge of `name` changes the situation in favor of "always visible" behavior.

thrix replied:
+1, thanks for looking up the details. Let's not block on this.

happz commented on tmt/steps/__init__.py:
```diff
@@ -62,6 +74,85 @@
+        '--update',
+        is_flag=True,
+        default=False,
+        help='Update existing phase. Use --name to specify which one.'),
```
Right, I'll improve the helm text.

happz replied:
Fixed in bbd6321b

thrix replied:
lgtm, thanks

thrix reviewed (approved):
No more comments from me.

You are welcome to merge @happz  ...

happz commented:
\o/ Thanks for the reviews, guys. One failed test was not related to the PR, and passed many times before.
